libs/barista-components/slider/src/slider.ts

```diff
@@ -0,0 +1,408 @@
+    }
+  }
+
+  /** Updates the input field with new value. */
+  private _updateInput(value: number): void {
+    this._sliderInput.value = value.toString();
+  }
+
+  /**
+   * Capture user events, and implement necessary streams to update view.
+   * TODO: refactor this to have cleaner streams, and possibly refactor into smaller functions.
+   */
+  private _captureEvents(): void {
+    const start$ = merge(
+      fromEvent(this._trackWrapper.nativeElement, 'mousedown'),
+      fromEvent(this._trackWrapper.nativeElement, 'touchstart'),
+    );
+
+    const move$ = merge(
+      fromEvent<MouseEvent>(window, 'mousemove').pipe(
+        map(mouseEvent => mouseEvent.clientX),
+      ),
+      fromEvent<TouchEvent>(window, 'touchmove').pipe(
+        map(touchEvent => touchEvent.changedTouches[0].clientX),
+      ),
+    );
+
+    const moveEnd$ = merge(
+      fromEvent(window, 'mouseup'),
+      fromEvent(window, 'touchend'),
+    );
+
+    /** Combining the start, move and moveEnd observers concludes the drag observer. */
+    const drag$ = start$.pipe(
+      switchMap(() =>
+        move$.pipe(
+          distinctUntilChanged(),
+          observeOn(animationFrameScheduler),
+          takeUntil(moveEnd$),
+        ),
+      ),
+      share(),
+    );
+
+    const click$ = fromEvent<MouseEvent>(
+      this._trackWrapper.nativeElement,
+      'click',
+    ).pipe(map(mouseEvent => mouseEvent.clientX));
+
+    // stream from keyboard events
+    const keyDown$ = fromEvent<KeyboardEvent>(
+      this._trackWrapper.nativeElement,
+      'keydown',
+    ).pipe(
+      filter(keyboardEvent => KEY_CODES_ARRAY.includes(keyboardEvent.keyCode)),
+      map(keyboardEvent => {
+        keyboardEvent.stopPropagation(); // global angular keydown would trigger CD.
+        const valueAddition = getKeyCodeValue(
+          this.max,
+          this.step,
+          keyboardEvent.keyCode,
+        );
+        const newValue = clamp(this._value + valueAddition, this.min, this.max);
+        return newValue;
+      }),
+      distinctUntilChanged(),
+      takeUntil(this._destroy$),
+    );
+
+    // triggered by drag and click
+    const mouse$ = merge(drag$, click$).pipe(
+      withLatestFrom(this._clientRect$),
+      map(([coordinate, { left, width }]) => {
+        const newValue = getSliderValueForCoordinate({
+          coordinate,
+          offset: left,
+          width,
+          min: this.min,
+          max: this.max,
+          step: this.step,
+          roundShift: this._roundShift,
+        });
+        return newValue;
+      }),
+      distinctUntilChanged(),
+      takeUntil(this._destroy$),
+    );
+
+    const inputValue$ = this.inputFieldValue$.pipe(
+      /**
+       * distinctUntilChanged() purposefully left out, to round the value
+       * and update the input field with the rounded value
+       */
+      map(value => roundToSnap(value, this.step, this.min, this.max)),
+      takeUntil(this._destroy$),
```

Comment:
please use the takeUntil always as the last operator before you subscribe.

Reply:
Thanks, will do.